cosmos-tx/tests/integration.rs

```diff
@@ -0,0 +1,191 @@
+            // There is a short period after the RPC health check succeeds
+            // and the `/genesis` endpoint returns valid data where
+            // transactions fail due to an internal SDK panic in the
+            // "undefined" code space:
+            //
+            //     TxResult {
+            //         code: Err(111222), data: None, log: Log("panic"), info: Info(""),
+            //         gas_wanted: Gas(0), gas_used: Gas(0), events: [],
+            //         codespace: Codespace("undefined")
+            //     }
+            //
+            // This should ideally get fixed upstream, however there is
+            // presently no tracking issue for this bug.
+            //
+            // If we see it pop up locally or in CI, perhaps we could find
+            // some way of polling readiness via RPC, or retrying when this
+            // specific error is encountered.
+            // TODO(tarcieri): open upstream issue and remove this hack when fixed
+            tokio::time::sleep(Duration::from_secs(2)).await;
+
+            let tx_commit_response = tx.broadcast_commit(&rpc_client).await.unwrap();
+
+            if tx_commit_response.check_tx.code.is_err() {
+                panic!("check_tx failed: {:?}", tx_commit_response.check_tx);
+            }
+
+            if tx_commit_response.deliver_tx.code.is_err() {
+                panic!("deliver_tx failed: {:?}", tx_commit_response.deliver_tx);
+            }
+
+            // TODO(tarcieri): look up transaction by hash and test transaction parsing
+        });
+    });
+}
+
+/// Initialize Tokio runtime
+fn init_tokio_runtime() -> tokio::runtime::Runtime {
+    tokio::runtime::Builder::new_current_thread()
+        .enable_all()
+        .build()
+        .unwrap()
+}
+
+/// Invoke `docker run` with the given arguments, calling the provided function
+/// after the container has booted and terminating the container after the
+/// provided function completes, catching panics and propagating them to ensure
+/// that the container reliably shuts down.
+///
+/// Prints log output from the container in the event an error occurred.
+fn docker_run<A, S, F>(args: A, f: F)
+where
+    A: IntoIterator<Item = S>,
+    S: AsRef<OsStr>,
+    F: FnOnce() -> () + panic::UnwindSafe,
+{
+    let container_id = exec_docker_command("run", args);
+    let result = panic::catch_unwind(f);
+
+    if result.is_err() {
+        let logs = exec_docker_command("logs", &[&container_id]);
+
+        println!("\n---- docker stdout ----");
+        println!("{}", logs);
+    }
+
+    exec_docker_command("kill", &[&container_id]);
+
+    if let Err(err) = result {
+        panic::resume_unwind(err);
+    }
+}
+
+/// Execute a given `docker` command, returning what was written to stdout
+/// if the command completed successfully.
+///
+/// Panics if the `docker` process exits with an error code
+fn exec_docker_command<A, S>(name: &str, args: A) -> String
```

Comment:
this is pretty cool.

Reply:
I was thinking of extracting this functionality in such a way it's reusable, e.g. behind a `dev` feature.

I think it'd be useful for testing TMKMS transaction signing.

But I thought I'd save that for a follow-up PR.